lib/class-wp-duotone-gutenberg.php

```diff
@@ -0,0 +1,366 @@
+	 * @return string The CSS variable.
+	 */
+	private static function get_css_var( $slug ) {
+		return 'var(' . self::get_css_custom_property_name( $slug ) . ')';
+	}
+
+	/**
+	 * Get the CSS declaration for a duotone preset.
+	 * Example: --wp--preset--duotone--blue-orange: url('#wp-duotone-blue-orange');
+	 *
+	 * @param array $filter_data The duotone data for presets and custom filters.
+	 * @return string The CSS declaration.
+	 */
+	private static function get_css_custom_property_declaration( $filter_data ) {
+		$declaration_value                = gutenberg_get_duotone_filter_property( $filter_data );
+		$duotone_preset_css_property_name = self::get_css_custom_property_name( $filter_data['slug'] );
+		return $duotone_preset_css_property_name . ': ' . $declaration_value . ';';
+	}
+
+	/**
+	 * Safari renders elements incorrectly on first paint when the SVG filter comes after the content that it is filtering,
+	 * so we force a repaint with a WebKit hack which solves the issue.
+	 *
+	 * @param string $selector The selector to apply the hack for.
+	 */
+	private static function safari_rerender_hack( $selector ) {
+		/*
+		* Simply accessing el.offsetHeight flushes layout and style
+		* changes in WebKit without having to wait for setTimeout.
+		*/
+		printf(
+			'<script>( function() { var el = document.querySelector( %s ); var display = el.style.display; el.style.display = "none"; el.offsetHeight; el.style.display = display; } )();</script>',
+			wp_json_encode( $selector )
+		);
+	}
+
+	/**
+	 * Outputs all necessary SVG for duotone filters, CSS for classic themes, and safari rerendering hack
+	 */
+	public static function output_footer_assets() {
+		foreach ( self::$output as $filter_data ) {
+
+			// SVG will be output on the page later.
+			$filter_svg = gutenberg_get_duotone_filter_svg( $filter_data );
+
+			echo $filter_svg;
+
+			// This is for classic themes - in block themes, the CSS is added in the head via wp_add_inline_style in the wp_enqueue_scripts action.
+			if ( ! wp_is_block_theme() ) {
+				wp_add_inline_style( 'core-block-supports', 'body{' . self::get_css_custom_property_declaration( $filter_data ) . '}' );
+			}
+
+			global $is_safari;
+			if ( $is_safari ) {
+				self::safari_rerender_hack( $filter_data['selector'] );
+			}
+		}
+	}
+
+	/**
+	 * Appends the used global style duotone filter CSS Vars to the inline global styles CSS
+	 */
+	public static function output_global_styles() {
+
+		if ( empty( self::$output ) ) {
+			return;
+		}
+
+		$duotone_css_vars = '';
+
+		foreach ( self::$output as $filter_data ) {
+			if ( ! array_key_exists( $filter_data['slug'], self::$global_styles_presets ) ) {
+				continue;
+			}
+
+			$duotone_css_vars .= self::get_css_custom_property_declaration( $filter_data );
+		}
+
+		if ( ! empty( $duotone_css_vars ) ) {
+			wp_add_inline_style( 'global-styles', 'body{' . $duotone_css_vars . '}' );
+		}
+	}
+
+	/**
+	 * Render out the duotone stylesheet and SVG.
```

Comment:
Tiny nit: since this is using the style engine to output styles within an existing stylesheet, should this refer to CSS styles instead of a separate stylesheet?

```suggestion
	 * Render out the duotone CSS styles and SVG.
```